src/transformers/generation_tf_logits_process.py

```
@@ -0,0 +1,299 @@
+        scores (`tf.Tensor` of shape `(batch_size, config.vocab_size)`):
+            Prediction scores of a language modeling head. These can be logits for each vocabulary when not using beam
+            search or log softmax for each vocabulary token when using beam search
+        kwargs:
+            Additional logits processor specific kwargs.
+
+    Return:
+        `tf.Tensor` of shape `(batch_size, config.vocab_size)`: The processed prediction scores.
+"""
+
+
+class TFLogitsProcessor(ABC):
+    """Abstract base class for all logit processors that can be applied during generation."""
+
+    @add_start_docstrings(TF_LOGITS_PROCESSOR_INPUTS_DOCSTRING)
+    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor) -> tf.Tensor:
+        """TF method for processing logits."""
+        raise NotImplementedError(
+            f"{self.__class__} is an abstract class. Only classes inheriting this class can be called."
+        )
+
+
+class TFLogitsProcessorList(list):
+    """
+    This class can be used to create a list of [`TFLogitsProcessor`] to subsequently process a `scores` input tensor.
+    This class inherits from list and adds a specific *__call__* method to apply each [`TFLogitsProcessor`] to the
+    inputs.
+    """
+
+    @add_start_docstrings(TF_LOGITS_PROCESSOR_INPUTS_DOCSTRING)
+    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, **kwargs) -> tf.Tensor:
+        for processor in self:
+            function_args = inspect.signature(processor.__call__).parameters
+            if len(function_args) > 3:
+                if not all(arg in kwargs for arg in list(function_args.keys())[2:]):
+                    raise ValueError(
+                        f"Make sure that all the required parameters: {list(function_args.keys())} for "
+                        f"{processor.__class__} are passed to the logits processor."
+                    )
+                scores = processor(input_ids, scores, **kwargs)
```

That's quite cool

---

Will this throw an error when a kwarg is only used by some of the `LogitsProcessor`s in the list? Most `LogitsProcessor` classes don't seem to have `**kwargs` in their call, so they will error out if they get an unexpected keyword argument, right?

---

It should throw an error if a logits processor requires more kwargs than the usual two (`input_ids` and `scores`) but it is not given to the list. Might be a bit overkill this check, but shouldn't hurt. You reminded me that the 3 is incorrect there though and that it should be a two :-)

---

I'm unsure though! Let's say we have two `LogitsProcessor`s in our list. `processor_1` takes `input_ids` and `scores` and `processor_2` takes `input_ids`, `scores` and `extra_data`. Then if we pass an `extra_data` argument to `TFLogitsProcessorList.__call__`, it will crash when it calls `processor_1` because `processor_1` has no `extra_data` or `**kwargs` argument. If we don't pass `extra_data` to `TFLogitsProcessorList.__call__` then it will crash when it calls `processor_2` because the `extra_data` argument for `processor_2` will be missing.

It seems like all `LogitsProcessor` classes need a `**kwargs` argument in their `__call__` method to catch extra unused arguments, or else they need to all have the same `__call__` signature.